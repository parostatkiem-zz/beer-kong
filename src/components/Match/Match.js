import React, { useState } from "react";
import { Jumbotron, Button } from "reactstrap";
import "./Match.scss";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faBolt,
  faTrophy,
  faFlagCheckered,
  faTrash
} from "@fortawesome/free-solid-svg-icons";
import { Link } from "react-router-dom";
import classNames from "classnames";
import useDate from "hooks/UseDate";
import EditMatchModal from "components/EditMatchModal/EditMatchModal";
import { useMutation } from "@apollo/react-hooks";
import { DELETE_MATCH } from "gql/mutations";
import { GET_LEAGUE } from "gql/queries";
import { GET_MATCHES } from "gql/queries";

function PlayerLink({ user }) {
  return (
    <Link to={"/player/" + user.id}>
      <img className="thumbnail" src={user.picture} alt={" "} />
      {user.name}
    </Link>
  );
}

const Match = ({
  user1,
  user2,
  plannedAt,
  winner = null,
  user1points,
  user2points,
  isFinished,
  leagueId,
  id,
  canManage = false
}) => {
  const date = useDate(plannedAt);
  const [deleteMatch] = useMutation(DELETE_MATCH, {
    refetchQueries: [
      { query: GET_LEAGUE, variables: { id: leagueId } },
      { query: GET_MATCHES, variables: { where: { league: { leagueId } } } }
    ],
    onError: e => window.alert(e.message),
    onCompleted: () => setShowEditModal(false)
  });

  function handleMatchDelete() {
    if (!window.confirm("Czy na pewno chcesz usunąć ten mecz?")) {
      return;
    }

    deleteMatch({ variables: { where: { id } } });
  }

  const [showEditModal, setShowEditModal] = useState(false);
  return (
    <>
      <Jumbotron className={classNames("match", { unplayed: !isFinished })}>
        <div
          className={classNames([
            "player",
            "left",
            { winner: winner && winner.id === user1.id }
          ])}
        >
          <span></span>
          <PlayerLink user={user1} />
          <span aria-label="user-one-points" className="points">
            {winner && winner.id === user1.id ? (
              <FontAwesomeIcon
                aria-label="trophy"
                className="trophy"
                icon={faTrophy}
              />
            ) : (
              <span></span>
            )}
            {user1points}
          </span>
        </div>
        <FontAwesomeIcon className="bolt_icon" icon={faBolt} />
        <span className="date">{date}</span>

        <div
          className={classNames([
            "player",
            "right",
            { winner: winner && winner.id === user2.id }
          ])}
        >
          <span aria-label="user-two-points" className="points">
            {user2points}
            {winner && winner.id === user2.id ? (
              <FontAwesomeIcon
                aria-label="trophy"
                className="trophy"
                icon={faTrophy}
              />
            ) : (
              <span></span>
            )}
          </span>
          <PlayerLink user={user2} />

          {canManage ? (
            <Button
              className="btn-icon btn-2 action-button"
              color="primary"
              type="button"
              size="sm"
              onClick={() =>
                isFinished ? handleMatchDelete() : setShowEditModal(true)
              }
            >
              <span className="btn-inner--icon">
                <FontAwesomeIcon
                  icon={isFinished ? faTrash : faFlagCheckered}
                />
              </span>
            </Button>
          ) : (
            <span></span>
          )}
        </div>
      </Jumbotron>
      <EditMatchModal
        isOpen={showEditModal}
        setOpen={setShowEditModal}
        wasFinished={isFinished}
        user1={user1}
        user2={user2}
        leagueId={leagueId}
        matchId={id}
        handleDelete={handleMatchDelete}
      />
    </>
  );
};

export default Match;
