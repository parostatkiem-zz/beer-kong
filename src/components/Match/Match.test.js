import React from "react";
import { render } from "@testing-library/react";
import Match from "./Match";
import { MockedProvider } from "@apollo/react-testing";
import { BrowserRouter } from "react-router-dom";

describe("Match", () => {
  const mockUsers = [
    {
      id: 1,
      name: "user1",
      picture: "user1picture"
    },
    {
      id: 2,
      name: "user2",
      picture: "user2picture"
    }
  ];
  it("Renders user names", () => {
    const { getByText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match user1={mockUsers[0]} user2={mockUsers[1]} />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(getByText(mockUsers[0].name)).toBeInTheDocument();
    expect(getByText(mockUsers[1].name)).toBeInTheDocument();
  });

  it("Renders user links with thumbnails", () => {
    const { queryAllByRole } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match user1={mockUsers[0]} user2={mockUsers[1]} />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(queryAllByRole("link")).toMatchSnapshot();
  });

  it("Renders user points", () => {
    const { getByLabelText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match
            user1={mockUsers[0]}
            user2={mockUsers[1]}
            user1points={3}
            user2points={4}
          />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(getByLabelText("user-one-points").textContent).toBe("3");
    expect(getByLabelText("user-two-points").textContent).toBe("4");
  });

  it("Renders trophy icon when the match is finished", () => {
    const { queryByLabelText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match winner={{ id: 2 }} user1={mockUsers[0]} user2={mockUsers[1]} />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(queryByLabelText("trophy")).toBeInTheDocument();
  });

  it("Does not render trophy icon when the match is not finished", () => {
    const { queryByLabelText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match isFinished={false} user1={mockUsers[0]} user2={mockUsers[1]} />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(queryByLabelText("trophy")).not.toBeInTheDocument();
  });

  it("Renders the planned date", () => {
    const mockDate = new Date("2006-11-05").toISOString();
    const { queryByText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Match
            plannedAt={mockDate}
            user1={mockUsers[0]}
            user2={mockUsers[1]}
          />
        </BrowserRouter>
      </MockedProvider>
    );
    expect(
      queryByText(new Date(mockDate).toLocaleDateString())
    ).toBeInTheDocument();
  });
});
