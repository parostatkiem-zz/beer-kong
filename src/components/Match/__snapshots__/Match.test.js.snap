// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match Renders user links with thumbnails 1`] = `
Array [
  <a
    href="/player/1"
  >
    <img
      alt=" "
      class="thumbnail"
      src="user1picture"
    />
    user1
  </a>,
  <a
    href="/player/2"
  >
    <img
      alt=" "
      class="thumbnail"
      src="user2picture"
    />
    user2
  </a>,
]
`;
